src/core/ipns/republisher.js

```diff
@@ -0,0 +1,124 @@
+
+const defaultBroadcastInterval = 4 * hour
+const defaultRecordLifetime = 24 * hour
+
+const ERR_NO_ENTRY_FOUND = 'ERR_NO_ENTRY_FOUND'
+const ERR_INVALID_IPNS_RECORD = 'ERR_INVALID_IPNS_RECORD'
+
+class IpnsRepublisher {
+  constructor (publisher, ipfs) {
+    this.publisher = publisher
+    this.ipfs = ipfs
+    this.repo = ipfs._repo
+  }
+
+  start () {
+    setInterval(() => {
+      this.republishEntries(this.ipfs._peerInfo.id.privKey, this.ipfs._options.pass)
+    }, defaultBroadcastInterval)
+  }
+
+  republishEntries (privateKey, pass) {
+    // TODO: Should use list of published entries.
+    // We can't currently *do* that because go uses this method for now.
+    this.republishEntry(privateKey, (err) => {
+      if (err) {
+        const error = 'cannot republish entry for the node\'s private key'
+
+        log.error(error)
+        return
+      }
+
+      if (this.ipfs._keychain && Boolean(pass)) {
+        this.ipfs._keychain.listKeys((err, list) => {
+          if (err) {
+            log.error(err)
+            return
+          }
+
+          each(list, (key, cb) => {
+            waterfall([
+              (cb) => this.ipfs._keychain.exportKey(key.name, pass, cb),
+              (pem, cb) => crypto.keys.import(pem, pass, cb)
+            ], (err, privKey) => {
+              if (err) {
+                log.error(err)
+                return
+              }
+
+              this.republishEntry(privKey, cb)
+            })
+          }, (err) => {
+            if (err) {
+              log.error(err)
+            }
+          })
+        })
+      }
+    })
+  }
+
+  republishEntry (privateKey, callback) {
+    waterfall([
+      (cb) => peerId.createFromPrivKey(privateKey.bytes.toString('base64'), cb),
+      (peerIdResult, cb) => this.getLastValue(peerIdResult, cb)
+    ], (err, value) => {
+      if (err) {
+        return callback(err.code === ERR_NO_ENTRY_FOUND ? null : err)
+      }
+
+      this.publisher.publishWithEOL(privateKey, value, defaultRecordLifetime, callback)
+    })
+  }
+
+  getLastValue (id, callback) {
+    this.repo.datastore.get(ipns.getLocalKey(id.id), (err, dsVal) => {
+      // error handling
+      // no need to republish
+      if (err && err.notFound) {
+        const error = `no previous entry for record with id: ${id}`
+
+        log.error(error)
+        return callback(Object.assign(new Error(error), { code: ERR_NO_ENTRY_FOUND }))
+      } else if (err) {
+        return callback(err)
+      }
+
+      if (!Buffer.isBuffer(dsVal)) {
+        const error = `found ipns record that we couldn't convert to a value`
+
+        log.error(error)
+        return callback(Object.assign(new Error(error), { code: ERR_INVALID_IPNS_RECORD }))
+      }
+
+      // unmarshal data
+      let record
+      try {
+        record = ipns.unmarshal(dsVal)
+      } catch (err) {
+        const error = `found ipns record that we couldn't convert to a value`
+
+        log.error(error)
+        return callback(error)
```

Comment:
`return callback(new Error(error))` - and I think that's why it's better called `errMsg` 😉